rfcs/2020-08-26-3191-host-metrics.md

```diff
@@ -0,0 +1,169 @@
+
+- https://crates.io/crates/heim (Rust)
+- https://lib.rs/crates/sys-info (Rust + C)
+- https://github.com/myfreeweb/systemstat (pure Rust)
+- https://docs.rs/sysinfo/0.3.19/sysinfo/index.html
+
+The Heim crate has a useful comparison doc showing the available tools and their capabilities.
+
+- https://github.com/heim-rs/heim/blob/master/COMPARISON.md
+
+For this implementation we're recommending the Heim crate based on platform and feature coverage.
+
+We'd use one of these to collect the following metrics:
+
+- `cpu_seconds_total` by mode (idle, nice, system, user) per CPU. (counter)
+- `disk_read_bytes_total` by disk (counter)
+- `disk_read_errors_total` by disk (counter)
+- `disk_read_retries_total` by disk (counter)
+- `disk_read_sectors_total` by disk (counter)
+- `disk_read_time_seconds_total` by disk (counter)
+- `disk_reads_completed_total` by disk (counter)
+- `disk_write_errors_total` by disk (counter)
+- `disk_write_retries_total` by disk (counter)
+- `disk_write_time_seconds_total` by disk (counter)
+- `disk_writes_completed_total` by disk (counter)
+- `disk_written_bytes_total` by disk (counter)
+- `disk_written_sectors_total` by disk (counter)
+- `filesystem_avail_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_device_error` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_total_file_nodes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_free_file_nodes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_free_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `filesystem_size_bytes` labeled with device, filesystem type, and mountpoint (gauge)
+- `load1` (gauge)
+- `load5` (gauge)
+- `load15` (gauge)
+- `memory_active_bytes` (gauge)
+- `memory_compressed_bytes` (gauge)
+- `memory_free_bytes` (gauge)
+- `memory_inactive_bytes` (gauge)
+- `memory_swap_total_bytes` (gauge)
+- `memory_swap_used_bytes` (gauge)
+- `memory_swapped_in_bytes_total` (gauge)
+- `memory_swapped_out_bytes_total` (gauge)
+- `memory_total_bytes` (gauge)
+- `memory_wired_bytes` (gauge)
+
+Users should be able to limit the collection of metrics to specific classes, here: `cpu`, `memory`, `disk/filesystem`, `load`.
+
+Metrics will also be labeled with:
+
+- `host`: the host name of the host being monitored.
+- `disk` for disk based metrics
+- `cpu` for CPU based metrics
+- `filesystem` for filesystem based metrics
+- `memory` for memory based metrics.
+- `load` for load based metrics.
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+[sources.my_source_id]
+  type = "host_metrics" # required
+  collecting = [ "all"] # optional, defaults collecting all metrics.
+  filesystem.mountpoints = [ "all" ] # optional, defaults to all mountpoints.
+  disk.devices = [ "all" ] # optional, defaults to all to disk devices.
```

Comment:
I'd recommend against magic words within the list to represent collecting all. We could:
1. not allow wildcard configuration like that, just say the default is all metrics;
2. use an actual wildcard to represent all, like `"*"`, which may be useful to allow for glob like `"filesystem_read*"`; or
3. use a custom serializer to allow either `"ALL"` outside of the list for all, or a list to enumerite items (ie `collecting = "ALL"` vs `collecting = [ "a", "b" ]`)

Reply:
I like the globbing. +1.